plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/packagejson/N4JSProjectSetupJsonValidatorExtension.xtend

```diff
@@ -403,23 +404,38 @@ public class N4JSProjectSetupJsonValidatorExtension extends AbstractJSONValidato
 		}
 		
 		if(!anyDependsOnTestLibrary(#[project])){
-			addIssuePreferred(#[], getMessageForSRCTEST_NO_TESTLIB_DEP(N4JSGlobals.MANGELHAFT), PROJECT_DEPENDENCY_CYCLE); 
+			addIssuePreferred(#[], getMessageForSRCTEST_NO_TESTLIB_DEP(N4JSGlobals.MANGELHAFT), SRCTEST_NO_TESTLIB_DEP);
 		}
 	}
 	
 	/**
 	 * check if any project in the list has dependency on test library, if so return true.
 	 * Otherwise invoke recursively in dependencies list of each project in initial list.
 	 *
-	 * NOTE: this implementation is not cycle safe!
-	 *
 	 * @returns true if any of the projects in the provided list depends (transitively) on the test library.
 	 */
 	private def boolean anyDependsOnTestLibrary(List<? extends IN4JSProject> projects) {
-		projects.findFirst[p|
-				p.dependencies.findFirst[N4JSGlobals.VENDOR_ID.equals(vendorID) && (N4JSGlobals.MANGELHAFT.equals(projectId) || N4JSGlobals.MANGELHAFT_ASSERT.equals(projectId))] !== null
-				|| anyDependsOnTestLibrary(p.dependencies)
-			] !== null
+		return anyDependsOnTestLibraryRecursive(projects, new HashSet<IN4JSProject>());
+	}
+
+	private def boolean anyDependsOnTestLibraryRecursive(List<? extends IN4JSProject> projects, Set<? extends IN4JSProject> visited) {
```

Comment:
needs recursion guard in case of cyclic dependency;
maybe `org.eclipse.n4js.utils.DependencyTraverser` could be used.